### PR TITLE
WIP: add code metrics

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

File: bugbug/utils.py [hunk 1]
```diff
@@ -309,7 +309,7 @@ def __setitem__(self, day, value):
 
 class LMDBDict:
     def __init__(self, path):
-        self.db = lmdb.open(path, map_size=68719476736, metasync=False, sync=False)
+        self.db = lmdb.open(path, map_size=68_719_476_736, metasync=False, sync=False)
```
Comment (marco-c) [inline bugbug/utils.py]: Don't change this unrelated file.

File: bugbug/models/regressor.py [hunk 2]
```diff
@@ -49,6 +49,7 @@ def __init__(self, lemmatization=False, interpretable=False):
 
         feature_extractors = [
             commit_features.source_code_file_size(),
+            commit_features.source_code_file_metrics(),
```
Comment (marco-c) [inline bugbug/models/regressor.py]: Don't add this yet to the two models, I want to add it in a separate PR so that we can measure the improvement in accuracy.

File: bugbug/repository.py [hunk 3]
```diff
@@ -77,9 +77,7 @@
     "Build System File": [".build", ".mk", ".in"],
 }
 
-HARDCODED_TYPES = {
-    ".eslintrc.js": ".eslintrc.js",
-}
+HARDCODED_TYPES = {".eslintrc.js": ".eslintrc.js"}
```
Comment (marco-c) [inline bugbug/repository.py]: Nit: don't change this, let black handle the formatting how it wants

File: bugbug/repository.py [hunk 4]
```diff
@@ -372,6 +378,23 @@ def _transform(commit):
 
             if size is not None:
                 source_code_sizes.append(size)
+                metrics = code_analysis_server.metrics(path, after)
+                if "error" in metrics:
```
Comment (marco-c) [inline bugbug/repository.py]: Why this if?

File: bugbug/repository.py [hunk 5]
```diff
@@ -372,6 +378,23 @@ def _transform(commit):
...
+                    ]
+                    commit.average_halstead_operands += metrics["halstead"]["operands"]
+                    commit.average_halstead_unique_operators += metrics["halstead"][
+                        "unique_operators"
+                    ]
+                    commit.average_halstead_operators += metrics["halstead"][
+                        "operators"
+                    ]
+                    commit.average_source_loc += metrics["loc"]["sloc"]
+                    commit.average_logical_loc += metrics["loc"]["lloc"]
```
Comment (marco-c) [inline bugbug/repository.py]: Are these at the file level or the function level?

File: bugbug/repository.py [hunk 6]
```diff
@@ -372,6 +378,23 @@ def _transform(commit):
...
+                        "unique_operands"
+                    ]
+                    commit.average_halstead_operands += metrics["halstead"]["operands"]
+                    commit.average_halstead_unique_operators += metrics["halstead"][
+                        "unique_operators"
+                    ]
+                    commit.average_halstead_operators += metrics["halstead"][
+                        "operators"
+                    ]
+                    commit.average_source_loc += metrics["loc"]["sloc"]
```
Comment (marco-c) [inline bugbug/repository.py]: Is this the number of lines of code, excluding comments?

File: bugbug/repository.py (hunk 5 above)
Comment (marco-c) [reply bugbug/repository.py]: They are at file level, as discussed on IRC, let's get mean/max/min/avg complexities both at function level and at file level

File: bugbug/repository.py (hunk 4 above)
Comment (calixteman) [reply bugbug/repository.py]: good catch: `not in` is better

File: bugbug/commit_features.py [hunk 7]
```diff
@@ -128,6 +128,47 @@ def __call__(self, commit, **kwargs):
         }
 
 
+class source_code_file_metrics(object):
+    name = "metrics on source code file"
+
+    def __call__(self, commit, **kwargs):
+        return {
```
Comment (marco-c) [inline bugbug/commit_features.py]: We could add the total (sum) too, like for other metrics

File: bugbug/repository.py [hunk 8]
```diff
@@ -378,6 +465,9 @@ def _transform(commit):
 
             if size is not None:
                 source_code_sizes.append(size)
+                metrics = code_analysis_server.metrics(path, after, unit=False)
```
Comment (marco-c) [inline bugbug/repository.py]: I forgot what the unit parameter is, could you document it in code_analysis_server?

File: bugbug/repository.py [hunk 9]
```diff
@@ -325,6 +411,7 @@ def _transform(commit):
     source_code_sizes = []
     other_sizes = []
     test_sizes = []
+    metrics_file_count = 0
```
Comment (marco-c) [inline bugbug/repository.py]: You are never increasing this :P

File: bugbug/repository.py [hunk 10]
```diff
@@ -415,6 +505,23 @@ def _transform(commit):
...
+    if metrics_file_count:
+        commit.average_cyclomatic /= metrics_file_count
+        commit.average_halstead_unique_operands /= metrics_file_count
+        commit.average_halstead_operands /= metrics_file_count
+        commit.average_halstead_unique_operators /= metrics_file_count
+        commit.average_halstead_operators /= metrics_file_count
+        commit.average_source_loc /= metrics_file_count
+        commit.average_logical_loc /= metrics_file_count
+    else:
+        commit.minimum_cyclomatic = 0
```
Comment (marco-c) [inline bugbug/repository.py]: Add a comment explaining that we need to do this when we didn't get any metrics and so the minimum would be sys.maxsize

File: bugbug/rust_code_analysis_server.py [hunk 11]
```diff
@@ -45,7 +45,7 @@ def start_process(self):
 
         try:
             self.proc = subprocess.Popen(
-                ["rust-code-analysis", "--serve", "--port", str(self.port),]
+                ["rust-code-analysis", "--serve", "--port", str(self.port)]
```
Comment (marco-c) [inline bugbug/rust_code_analysis_server.py]: WTH, why is black giving different results for you and me? :P

File: bugbug/commit_features.py [hunk 12]
```diff
@@ -128,6 +128,54 @@ def __call__(self, commit, **kwargs):
...
+                "minimum_halstead_unique_operands"
+            ],
+            "Minimum number of operands": commit["minimum_halstead_operands"],
+            "Minimum number of unique operators": commit[
+                "minimum_halstead_unique_operators"
+            ],
+            "Minimum number of operators": commit["minimum_halstead_operators"],
+            "Minimum number of source loc": commit["minimum_source_loc"],
+            "Minimum number of logical loc": commit["minimum_logical_loc"],
+            "Sum of number of operands": commit["sum_halstead_operands"],
```
Comment (marco-c) [inline bugbug/commit_features.py]: Use `Total` instead of `Sum of`, as there is some code which simplifies the names and relies on `Maximum`, `Minimum`, `Total` and `Average` being there

File: bugbug/rust_code_analysis_server.py (hunk 11 above)
Comment (calixteman) [reply bugbug/rust_code_analysis_server.py]: I don't know... my editor was open before black ran and the comma is there... so it removed it

File: bugbug/repository.py (hunk 6 above)
Comment (calixteman) [reply bugbug/repository.py]: sloc is just function span length and lloc is the number of line with something of interest.
For example, for c++:
https://github.com/mozilla/rust-code-analysis/blob/master/src/sloc.rs#L164-L178
